panda/src/navigation/navMesh.cxx

```diff
@@ -0,0 +1,388 @@
+  _params.detailVerts = mesh_params.detail_verts;
+  _params.detailVertsCount = mesh_params.detail_vert_count;
+  _params.detailTris = mesh_params.detail_tris;
+  _params.detailTriCount = mesh_params.detail_tri_count;
+
+
+  _params.walkableHeight = mesh_params.walkable_height;
+  _params.walkableRadius = mesh_params.walkable_radius;
+  _params.walkableClimb = mesh_params.walkable_climb;
+
+  _params.bmin[0] = mesh_params.b_min[0];
+  _params.bmin[1] = mesh_params.b_min[1];
+  _params.bmin[2] = mesh_params.b_min[2];
+  _params.bmax[0] = mesh_params.b_max[0];
+  _params.bmax[1] = mesh_params.b_max[1];
+  _params.bmax[2] = mesh_params.b_max[2];
+  
+  _params.cs = mesh_params.cs;
+  _params.ch = mesh_params.ch;
+  _params.buildBvTree = mesh_params.build_bv_tree;
+  
+  RC_MESH_NULL_IDX = mesh_params.RC_MESH_NULL_IDX;
+  
+  init_nav_mesh();
+}
+
+bool NavMesh::init_nav_mesh() {
+	unsigned char *nav_data = 0;
+  int nav_data_size = 0;
+
+  if (!dtCreateNavMeshData(&_params, &nav_data, &nav_data_size)) {
+    std::cout<<"\nCould not build Detour navmesh.\n";
+    return false;
+  }
+
+  _nav_mesh = dtAllocNavMesh();
+
+  if (!_nav_mesh) {
+    dtFree(nav_data);
+    std::cout<<"\nCould not create Detour navmesh\n";
+    return false;
+  }
+
+  dtStatus status;
+
+  status = _nav_mesh->init(nav_data, nav_data_size, DT_TILE_FREE_DATA);
+  if (dtStatusFailed(status)) {
+    dtFree(nav_data);
+    std::cout<<"\nCould not init Detour navmesh\n";
+    return false;
+  }
+}
+
+
+PT(GeomNode) NavMesh::draw_nav_mesh_geom() {
+
+  PT(GeomVertexData) vdata;
+  vdata = new GeomVertexData("vertexInfo", GeomVertexFormat::get_v3c4(), Geom::UH_static);
+  std::cout << "\nJust degbugging for navMesh: navmesh vert count: " << _params.vertCount << "\n";
+  vdata->set_num_rows(_params.vertCount);
+
+  GeomVertexWriter vertex(vdata, "vertex");
+  GeomVertexWriter colour(vdata, "color");
+
+  const int nvp = _params.nvp;
+  std::cout << "nvp: " << nvp << std::endl;
+  const float cs = _params.cs;
+  std::cout << "cs: " << cs << std::endl;
+  const float ch = _params.ch;
+  std::cout << "ch: " << ch << std::endl;
+  const float* orig = _params.bmin;
+  std::cout << "orig: " << orig[0] << "\t" << orig[1] << "\t" << orig[2] << std::endl;
+
+  std::cout << "NavMesh poly count: " << _params.polyCount << std::endl;
+  std::cout << "NavMesh vert count: " << _params.vertCount << std::endl;
+
+  for (int i = 0;i < _params.vertCount * 3;i += 3) {
+
+    const unsigned short* v = &_params.verts[i];
+
+    //convert to world space
+    const float x = orig[0] + v[0] * cs;
+    const float y = orig[1] + v[1] * ch;
+    const float z = orig[2] + v[2] * cs;
+
+    vertex.add_data3(x, -z, y); //if origingally model is z-up
+    //vertex.add_data3(x, y, z); //if originally model is y-up
+    colour.add_data4((float)rand() / RAND_MAX, (float)rand() / RAND_MAX, (float)rand() / RAND_MAX, 1);
+    //std::cout << "index: " << i / 3 << "\t" << x << "\t" << y << "\t" << z << "\n";
```

Comment:
If there is time you should take into account that the `coordinate-system` Config.prc variable may not be set to Z-up necessarily.  It may be necessary to, before the loop, store something like:
```python
LMatrix4 mat = LMatrix4::convert_mat(CS_yup_right, CS_default);

And then during the loop use `mat.convert_point` to convert your points according to the given transformation.

Either that, or you can pass the matrix to a call to `geom.transform_vertices` after the fact.
```

Reply:
Done